File: android/src/main/java/app/capgo/plugin/health/background/BackgroundHealthCoordinator.kt
package app.capgo.plugin.health.background

import android.content.Context
import android.util.Log
import androidx.health.connect.client.HealthConnectClient
import androidx.work.ListenableWorker
import app.capgo.plugin.health.HealthDataType
import app.capgo.plugin.health.HealthManager
import com.getcapacitor.JSArray
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

class BackgroundHealthCoordinator(
    private val context: Context,
    private val preferences: BackgroundHealthPreferences = BackgroundHealthPreferences(context),
    private val healthManager: HealthManager = HealthManager(),
    private val permissionChecker: BackgroundHealthPermissionChecker =
        BackgroundHealthPermissionChecker(context, healthManager),
    private val apiClient: BackgroundHealthApiClient = BackgroundHealthApiClient()
) {
    suspend fun run(): ListenableWorker.Result {
        val config = preferences.getConfig() ?: return ListenableWorker.Result.success()
        if (!config.enabled) {
            return ListenableWorker.Result.success()
        }

        val status = HealthConnectClient.getSdkStatus(context)
        if (status != HealthConnectClient.SDK_AVAILABLE) {
            return ListenableWorker.Result.retry()
        }

        val client = HealthConnectClient.getOrCreate(context)
        if (!permissionChecker.hasAllPermissionsForBackgroundSync(client, config)) {
            return ListenableWorker.Result.retry()
        }

        val lastSyncMap = try {
            apiClient.fetchLastSyncMap(config.getLastSync, config.subjectId)
        } catch (error: Exception) {
            Log.w(TAG, "Background sync failed fetching last-sync state.", error)
            return ListenableWorker.Result.retry()
        }

        val uploadedSamples = JSArray()
        var successfulReadCount = 0

        config.dataTypes.forEach { dataType ->
            try {
                val window = resolveReadWindow(lastSyncMap, dataType)
                val samples = healthManager.readSamples(
                    client = client,
                    dataType = dataType,
                    startTime = window.start,
                    endTime = window.end,
                    limit = 0,
                    ascending = true
                )
                for (index in 0 until samples.length()) {
                    uploadedSamples.put(samples.opt(index))
                }
                successfulReadCount += 1
            } catch (error: Exception) {
                Log.w(TAG, "Background sync read failed for ${dataType.identifier}. Continuing with partial upload.", error)
            }
        }

        if (successfulReadCount == 0) {
            return ListenableWorker.Result.retry()
        }

        if (uploadedSamples.length() == 0) {
            // POST /api/health/signals requires non-empty `data`; nothing new from Health Connect this run.
            return ListenableWorker.Result.success()
        }

        return try {
            apiClient.uploadSamples(config.postSamples, config.subjectId, uploadedSamples)
            ListenableWorker.Result.success()
        } catch (error: Exception) {
            Log.w(TAG, "Background sync upload failed.", error)
            ListenableWorker.Result.retry()
        }
    }

    /**
     * Resolves the Health Connect read interval for one [dataType].
     *
     * Uses [Instant.now] at call time as the upper bound where applicable.
     *
     * 1. No last-sync entry: start = local midnight today, end = now.
     * 2. Last sync more than 24h before now: start = stored instant, end = that instant + 24h (backfill chunk).
     * 3. Otherwise: start = stored instant, end = now.
     */
    private fun resolveReadWindow(
        lastSyncMap: Map<HealthDataType, String>,
        dataType: HealthDataType
    ): ReadWindow {
        val now = Instant.now()
        val raw = lastSyncMap[dataType]
        if (raw == null) {
            val zone = ZoneId.systemDefault()
            val startOfToday = LocalDate.now(zone).atStartOfDay(zone).toInstant()
            return ReadWindow(start = startOfToday, end = now)
        }

        val lastSyncTimestamp = try {
            Instant.parse(raw)
        } catch (error: Exception) {
            throw IllegalArgumentException("Invalid ISO timestamp for ${dataType.identifier}: $raw", error)
        }

        val cutoff = now.minus(MAX_WINDOW)
        return if (lastSyncTimestamp.isBefore(cutoff)) {
            ReadWindow(start = lastSyncTimestamp, end = lastSyncTimestamp.plus(MAX_WINDOW))
        } else {
            ReadWindow(start = lastSyncTimestamp, end = now)
        }
    }

    private data class ReadWindow(val start: Instant, val end: Instant)

    companion object {
        private val MAX_WINDOW: Duration = Duration.ofHours(24)
        private const val TAG = "BackgroundHealthSync"
    }
}

File: ios/Sources/HealthPlugin/background/BackgroundHealthCoordinator.swift
import Foundation
import HealthKit

enum BackgroundHealthRunOutcome {
    case success
    case retry
}

final class BackgroundHealthCoordinator {
    private let preferences: BackgroundHealthPreferences
    private let apiClient: BackgroundHealthApiClient
    private let isoFormatter: ISO8601DateFormatter
    private static let maxWindow: TimeInterval = 24 * 60 * 60

    init(
        preferences: BackgroundHealthPreferences = BackgroundHealthPreferences(),
        apiClient: BackgroundHealthApiClient = BackgroundHealthApiClient()
    ) {
        self.preferences = preferences
        self.apiClient = apiClient
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        isoFormatter = formatter
    }

    func run(health: Health, completion: @escaping (BackgroundHealthRunOutcome) -> Void) {
        guard let config = preferences.getConfig(), config.enabled else {
            completion(.success)
            return
        }

        guard HKHealthStore.isHealthDataAvailable() else {
            completion(.retry)
            return
        }

        apiClient.fetchLastSyncMap(config: config.getLastSync, subjectId: config.subjectId) { [weak self] result in
            guard let self = self else { return }
            switch result {
            case let .failure(error):
                NSLog("[BackgroundHealthSync] Failed fetching last-sync state: \(error)")
                completion(.retry)
            case let .success(lastSyncMap):
                self.readAndUpload(config: config, health: health, lastSyncMap: lastSyncMap, completion: completion)
            }
        }
    }

    private func readAndUpload(
        config: BackgroundSyncConfig,
        health: Health,
        lastSyncMap: [String: String],
        completion: @escaping (BackgroundHealthRunOutcome) -> Void
    ) {
        let group = DispatchGroup()
        let lock = NSLock()
        var uploadedSamples: [[String: Any]] = []
        var successfulReadCount = 0

        for identifier in config.dataTypes {
            let window = resolveReadWindow(lastSyncMap: lastSyncMap, dataTypeIdentifier: identifier)
            group.enter()
            do {
                try health.readSamples(
                    dataTypeIdentifier: identifier,
                    startDateString: isoFormatter.string(from: window.start),
                    endDateString: isoFormatter.string(from: window.end),
                    limit: 0,
                    ascending: true
                ) { result in
                    defer { group.leave() }
                    switch result {
                    case let .success(samples):
                        lock.lock()
                        successfulReadCount += 1
                        uploadedSamples.append(contentsOf: samples)
                        lock.unlock()
                    case let .failure(error):
                        NSLog(
                            "[BackgroundHealthSync] Read failed for \(identifier). Continuing with partial upload. \(error)"
                        )
                    }
                }
            } catch {
                NSLog("[BackgroundHealthSync] Read setup failed for \(identifier): \(error)")
                group.leave()
            }
        }

        group.notify(queue: .global(qos: .utility)) {
            if successfulReadCount == 0 {
                completion(.retry)
                return
            }
            if uploadedSamples.isEmpty {
                completion(.success)
                return
            }

            self.apiClient.uploadSamples(
                config: config.postSamples,
                subjectId: config.subjectId,
                samples: uploadedSamples
            ) { uploadResult in
                switch uploadResult {
                case .success:
                    completion(.success)
                case let .failure(error):
                    NSLog("[BackgroundHealthSync] Upload failed: \(error)")
                    completion(.retry)
                }
            }
        }
    }

    private func resolveReadWindow(lastSyncMap: [String: String], dataTypeIdentifier: String) -> (start: Date, end: Date) {
        let now = Date()
        guard let raw = lastSyncMap[dataTypeIdentifier] else {
            let startOfToday = Calendar.current.startOfDay(for: now)
            return (startOfToday, now)
        }

        guard let lastSync = isoFormatter.date(from: raw) ?? ISO8601DateFormatter().date(from: raw) else {
            let startOfToday = Calendar.current.startOfDay(for: now)
            return (startOfToday, now)
        }

        let cutoff = now.addingTimeInterval(-Self.maxWindow)
        if lastSync < cutoff {
            return (lastSync, lastSync.addingTimeInterval(Self.maxWindow))
        }
        return (lastSync, now)
    }
}
